### DISCUSSION TITLE
How to trigger center action by passing position/coordinates as parameter

### DISCUSSION THREAD

**Discussion by sathya-1994** [eclipse-glsp/glsp#746] [Q&A] [answered]
I am using vscode version of GLSP,

Currently we can use org.eclipse.glsp.server.actions.CenterAction to bring any element to center of the viewport by passing ID as parameter.

But I want pass extra parameter along with ID so that particular region/part of element will be shown at the center of viewport.

example : 
1. coordinate (10,10) which will bring relative position of node to center of the viewport
2. Top Left Corner : Top left corner of node will be placed at the center of viewport

Is any way we can implement this requirement in GLSP?

**Comment by planger** [comment, marked as answer]
Yes, you can create an own action that extends `CenterAction` and register a custom command for your new action that takes a relative offset into account. So best look at the implementation of the `CenterAction` and adjust to your needs:

* [Center Action](https://github.com/eclipse/sprotty/blob/b0e9d60e9c9ab7d302f84f55eef227828ba7d68c/packages/sprotty/src/features/viewport/center-fit.ts#L41)
* [Registration](https://github.com/eclipse/sprotty/blob/b0e9d60e9c9ab7d302f84f55eef227828ba7d68c/packages/sprotty/src/features/viewport/di.config.ts#L26) of the existing [center command](https://github.com/eclipse/sprotty/blob/b0e9d60e9c9ab7d302f84f55eef227828ba7d68c/packages/sprotty/src/features/viewport/center-fit.ts#L165)

So you would create a new subclass of `CenterAction` and a subclass of `CenterCommand` (or `BoundsAwareViewportCommand`), register the command as linked above in your `di.config.ts` and adjust the implementation of your `CenterCommand`, e.g. as follows (see `this.action.yourRelativeOffsetX` and Y):

```typescript
    getNewViewport(bounds: Bounds, model: SModelRoot): Viewport | undefined {
        if (!Dimension.isValid(model.canvasBounds)) {
            return undefined;
        }
        const zoom = (this.action.retainZoom && isViewport(model)) ? model.zoom : 1;
        const c = Bounds.center(bounds);
        return {
            scroll: {
                x: c.x + this.action.yourRelativeOffsetX - 0.5 * model.canvasBounds.width / zoom,
                y: c.y + this.action.yourRelativeOffsetY - 0.5 * model.canvasBounds.height / zoom
            },
            zoom: zoom
        };
    }
```

Alternatively, you can add a child node to your node and place it somehwere within its parent node and then center child node instead of the parent node. This may even be the better solution, depending on the structure of your `GModel`.
